### PR TITLE
[MPS] Add C++ API support for MPS backend

### LINKED ISSUE #96425: Expose MPS to C++ (like cuda)
### 🚀 The feature, motivation and pitch

There is torch::cuda::is_available(), but there is no torch::mps::is_available(), so it looks like mps/metal is not exposed as a backend to C++.

Some apps might want to run inference on metal using the very nice to use pytorch apis + it would be very helpful to be able to test cuda like flows (like sending tensors to gpu etc...) locally on a mac by using MPS (Metal).

### Alternatives

_No response_

### Additional context

_No response_

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by razarmehr**
- This enables the APIs `torch::mps::is_available()/synchronize()/manual_seed()` for use in PyTorch C++.
- Added test case for C++ APIs to `mps_test_allocator.cpp`

Fixes #96425

### CODE REVIEW

**Review by kulinseth — APPROVED** [review]
Looks good to me.  @albanD please take a look

File: aten/src/ATen/test/mps_test_allocator.cpp [hunk 1]
```diff
@@ -35,5 +38,6 @@ TEST(MPSAllocator, MPSAllocatorCallbacks) {
         }
         replay_buffer.push_back(new_value);
     }
+    torch::mps::synchronize();
```
Comment (albanD) [inline aten/src/ATen/test/mps_test_allocator.cpp:41]: @kulinseth Is there doc about what this op does? Do we actually need to call it here?
If this is just to check the function is available, we should have a comment saying so.

**Review by albanD — COMMENTED** [review]
Change sounds ok. nit on the test

File: aten/src/ATen/test/mps_test_allocator.cpp (hunk 1 above)
Comment (kulinseth) [reply aten/src/ATen/test/mps_test_allocator.cpp:41]: @albanD , the synchronize is documented [here ](https://github.com/pytorch/pytorch/pull/96668/files#diff-46701bb5876c77c75fc9577322d64f2f41daa5112d8ac83c3da9641679b1d854R17). It's just a way for MPS to synchronize with CPU, similar to CUDA synchronize function. Do you think there is value to add this in the MPS docs ?

Comment (razarmehr) [reply aten/src/ATen/test/mps_test_allocator.cpp:41]: Yes, the `torch.mps.synchronize()` is documented [in here](https://github.com/pytorch/pytorch/pull/94417/files#diff-3a9cc21de8cd29d75bfce4d86e4ebdbe8a68fb937b8a7b972dcc7f73d7e710d1R17) and compiled via `mps.rst`. So it's documentation should be available on on torch-2.0.

Comment (razarmehr) [reply aten/src/ATen/test/mps_test_allocator.cpp:44]: I added a comment for that line anyway. Thanks.

Comment (albanD) [reply aten/src/ATen/test/mps_test_allocator.cpp:44]: Yeah I think it is an ok description. But my guess is that people may wonder when they need to call this or not.
Not that the cuda version does much better. But I think we should mention that we don't expect the user to ever call this unless they're doing benchmarking on need a full sync with the GPU for some reason.

**Review by albanD — APPROVED** [review]
Thanks!

File: torch/csrc/api/include/torch/mps.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+bool TORCH_API is_available();
+
+/// Sets the RNG seed for the MPS device.
+void TORCH_API manual_seed(uint64_t seed);
+
+/// Waits for all streams on a MPS device to complete.
+void TORCH_API synchronize();
+
+} // namespace mps
+} // namespace torch
```
Comment (DanilBaibak) [inline torch/csrc/api/include/torch/mps.h:21]: Please add a blank line to the end of the file.

File: torch/csrc/api/src/mps.cpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  }
+}
+
+void synchronize() {
+  TORCH_CHECK(is_available(), "No MPS devices are available");
+  at::detail::getMPSHooks().deviceSynchronize();
+}
+
+} // namespace mps
+} // namespace torch
```
Comment (DanilBaibak) [inline torch/csrc/api/src/mps.cpp:33]: Please add a blank line to the end of the file.

Comment (kulinseth) [reply torch/csrc/api/src/mps.cpp:33]: Hi @DanilBaibak , how can we enable this build in the pr ?

**Comment by razarmehr** [comment]
@DanilBaibak , sure will do. Interesting that I added `ciflow/trunk` to labels and used regular merge and none of the checks failed. What's the CI flow to catch such cosmetic errors?

**Comment by DanilBaibak** [comment]
@razarmehr, there is a discrepancy between GitHub and local build. We are working to resolve this inconsistency. Sorry for the inconvenience!

File: torch/csrc/api/src/mps.cpp (hunk 3 above)
Comment (razarmehr) [reply torch/csrc/api/src/mps.cpp]: Added

File: torch/csrc/api/include/torch/mps.h (hunk 2 above)
Comment (razarmehr) [reply torch/csrc/api/include/torch/mps.h]: Added

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Hmm, I remember it was agreed that internal linters are disabled for PyTorch codebase,  so this change should not have been reverted

**Pull request closed without merging**